### PR TITLE
add data service which calls servlets

### PR DESCRIPTION
**Pull request by adiassulin**

### CODE REVIEW

**Comment by chenraff** [comment]
LGTM 👍

**Review by chenraff — APPROVED** [review]
LGTM 👍

File: trendly/src/app/histogram-section/histogram-section.component.ts [hunk 1]
```diff
@@ -57,14 +58,14 @@ export class HistogramSectionComponent implements OnInit {
 
   ngOnInit(): void {
     // this function should be call after data retrival from server, will change after creating backend
-    this.convertDataToChartsFormat();
+    // this.convertDataToChartsFormat();
```
Comment (mayascha) [inline trendly/src/app/histogram-section/histogram-section.component.ts:61]: Is this intended?

File: trendly/src/app/data.service.ts [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    const termParameter : string = this.makeTermParameter(term);
+    const dateParameter : string = START_DATE_PARAMETER + startDate + NEXT_PARAMETER + END_DATE_PARAMETER + endDate + NEXT_PARAMETER;
+    const countryParameter : string = COUNTRY_PARAMETER + country + NEXT_PARAMETER;
+    const intervalParameter : string = INTERVAL_PARAMETER + interval;
+
+    return termParameter + dateParameter + countryParameter + intervalParameter;
+  }
+
+  /**
+   * returns the term parameter part of the url.
```
Comment (mayascha) [inline trendly/src/app/data.service.ts:59]: Copital letters in the beginning of comments

File: trendly/src/app/histogram-section/histogram-section.component.ts [hunk 3]
```diff
@@ -132,5 +135,14 @@ export class HistogramSectionComponent implements OnInit {
       array[i] = 0;
     }
 
-  } 
+  }
+  
+    ngOnChanges(changes: SimpleChanges): void {
+    if(changes['trendsData'])
+    {
```
Comment (mayascha) [inline trendly/src/app/histogram-section/histogram-section.component.ts:142]: new line is not needed for {

File: trendly/src/app/histogram-section/histogram-section.component.ts [hunk 4]
```diff
@@ -107,12 +109,13 @@ export class HistogramSectionComponent implements OnInit {
     * @param topics
     */
    private createData(topics : Map<string, number>) : void{
-     for (let date of Object.keys(MOCK_DATA)) {
+     this.data = [];
+     for (let date of Object.keys(this.trendsData)) {
        const row = Array((topics.size * 3) + 1).fill('');
        this.initializeRowArray(row);
```
Comment (mayascha) [inline trendly/src/app/histogram-section/histogram-section.component.ts:115]: Looks like 3 is a magic number in a lot of your code. worth giving it a meaningful constant name

File: trendly/src/app/data.service.ts [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+   * @param term 
+   */
+  private makeTermParameter(term: string | string[]) : string {
```
Comment (mayascha) [inline trendly/src/app/data.service.ts:62]: replace term with termOrTerms

File: trendly/src/app/data.service.ts [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+    if (typeof term === TYPE_STRING) {
+      return TERM_PARAMETER + term + NEXT_PARAMETER;
+    }
+
+    let terms : string = EMPTY_STR;
+    for (let str of term) {
+      terms += TERM_PARAMETER + str + NEXT_PARAMETER;
+    }
+    return terms;
```
Comment (mayascha) [inline trendly/src/app/data.service.ts:71]: return termOrTerms.reduce((terms, term) => terms + TERM_PARAMETER + term + NEXT_PARAMETER, EMPTY_STR);

File: trendly/src/app/data.service.ts [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+  }
+
+  /**
+   * returns the term parameter part of the url.
+   * @param term 
+   */
+  private makeTermParameter(term: string | string[]) : string {
+    if (typeof term === TYPE_STRING) {
+      return TERM_PARAMETER + term + NEXT_PARAMETER;
+    }
```
Comment (mayascha) [inline trendly/src/app/data.service.ts:65]: Can be in 1 line without {}

File: trendly/src/app/data.service.ts [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+    return termParameter + dateParameter + countryParameter + intervalParameter;
+  }
+
+  /**
+   * returns the term parameter part of the url.
+   * @param term 
+   */
+  private makeTermParameter(term: string | string[]) : string {
+    if (typeof term === TYPE_STRING) {
+      return TERM_PARAMETER + term + NEXT_PARAMETER;
```
Comment (mayascha) [inline trendly/src/app/data.service.ts:64]: create lambda and usu is bellow
const wrapTerm = (term) => TERM_PARAMETER + term + NEXT_PARAMETER;

File: trendly/src/app/data.service.ts [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+    }
+
+    let terms : string = EMPTY_STR;
+    for (let str of term) {
+      terms += TERM_PARAMETER + str + NEXT_PARAMETER;
+    }
+    return terms;
+  }
+}
```
Comment (mayascha) [inline trendly/src/app/data.service.ts:73]: the entire func can actually be 

const wrapTerm = (term) => TERM_PARAMETER + term + NEXT_PARAMETER;
return typeof termOrTerms === TYPE_STRING ? wrapTerm(termOrTerms) : termOrTerms.map((term) => wrapTerm(term)).join(EMPTY_STR);

File: trendly/src/app/histogram-section/histogram-section.component.ts (hunk 1 above)
Comment (adiassulin) [reply trendly/src/app/histogram-section/histogram-section.component.ts:61]: for initial view I was wondering whether to call it from ngOnInit or from the constructor. for all later histogram vies the function called from the ngOnChane function.

File: trendly/src/app/data.service.ts (hunk 2 above)
Comment (adiassulin) [reply trendly/src/app/data.service.ts:59]: done

File: trendly/src/app/histogram-section/histogram-section.component.ts (hunk 3 above)
Comment (adiassulin) [reply trendly/src/app/histogram-section/histogram-section.component.ts:142]: done

File: trendly/src/app/histogram-section/histogram-section.component.ts (hunk 4 above)
Comment (adiassulin) [reply trendly/src/app/histogram-section/histogram-section.component.ts:115]: done

File: trendly/src/app/data.service.ts (hunk 9 above)
Comment (adiassulin) [reply trendly/src/app/data.service.ts:73]: looks good, changed it

File: trendly/src/app/histogram-section/histogram-section.component.ts [hunk 10]
```diff
@@ -1,136 +1,157 @@
...
     isStacked: true,
     colors: this.coloresService._lightColorShow,
   };
 
   /**
-   * the constructor - injects color service.
-   * @param coloresService 
+   * The constructor - injects color service.
+   * @param coloresService
    */
```
Comment (ewebdev) [inline trendly/src/app/histogram-section/histogram-section.component.ts:68]: Redundant, remove.

File: trendly/src/app/histogram-section/histogram-section.component.ts [hunk 11]
```diff
@@ -1,136 +1,157 @@
...
-  constructor(private coloresService : ColorsService) {}
+  constructor(private coloresService: ColorsService) {}
 
   ngOnInit(): void {
-    // this function should be call after data retrival from server, will change after creating backend
-    this.convertDataToChartsFormat();
+    // this function should be call after data retrival from server, will change
+    // after creating backend this.convertDataToChartsFormat();
```
Comment (ewebdev) [inline trendly/src/app/histogram-section/histogram-section.component.ts:73]:
```suggestion
    // TODO: this function should be call after data retrieval from server, will change
    // after creating backend this.convertDataToChartsFormat();
```

File: trendly/src/app/histogramy-component/histogramy-component.component.ts [hunk 12]
```diff
@@ -1,4 +1,4 @@
-import { Component, OnInit } from '@angular/core';
+import { Component, OnInit, Input } from '@angular/core';
```
Comment (ewebdev) [inline trendly/src/app/histogramy-component/histogramy-component.component.ts:1]: Formatting

File: trendly/src/app/histogram-section/histogram-section.component.ts [hunk 13]
```diff
@@ -1,136 +1,157 @@
...
+  createColumnNames(topics: Map<string, number>): void {
+    this.columnNames = [];
+    this.columnNames[0] = 'Topic';
+    let index: number = 0;
+    for (let key of topics.keys()) {
+      index = topics.get(key) * 3;
+      this.columnNames[index + 1] = key;
+      this.columnNames[index + 2] = {role: TOOLTIP_ROLE};
+      this.columnNames[index + 3] = {role: STYLE_ROLE};
+    }
```
Comment (ewebdev) [inline trendly/src/app/histogram-section/histogram-section.component.ts:115]: Tip: use Map.entries() together with destructuring assignment
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment
as follows:
```suggestion
    for (const [key, i] of topics.entries()) {
      let index = i*3;
      this.columnNames[++index] = key;
      this.columnNames[++index] = {role: TOOLTIP_ROLE};
      this.columnNames[++index] = {role: STYLE_ROLE};
    }
```

File: trendly/src/app/histogramy-component/histogramy-component.component.ts [hunk 14]
```diff
@@ -20,22 +20,33 @@ interface InputObj {
...
   readonly TopTopicsTitle : string = 'Top Topics'
   readonly RisingTopicsTitle : string = 'Rising Topics'
+  dataTop;
+  dataRising;
 
   constructor(private dataService : DataService) { }
 
   ngOnInit(): void {
-    console.log(this.dataService.callServlet('/trendly', ['t', 'y', 'l'], '2016 2', '2017 7', 'US', 3));
   }
```
Comment (ewebdev) [inline trendly/src/app/histogramy-component/histogramy-component.component.ts:29]: Remove

File: trendly/src/app/histogramy-component/histogramy-component.component.ts [hunk 15]
```diff
@@ -20,22 +20,33 @@ interface InputObj {
...
    * gets data from server according to the given parameters. (for now just printing).
-   * @param input 
+   * @param input
    */
   getDataFromServer(input: InputObj) {
     console.log(input['term']);
     console.log(input['interval']);
     console.log(input['startDate']);
     console.log(input['endDate']);
     console.log(input['country']);
```
Comment (ewebdev) [inline trendly/src/app/histogramy-component/histogramy-component.component.ts:40]: Tip: can print all values in a single console.log row as follows:
```
console.log(input);
```
or
```
console.log({term: input['term'], interval: input['interval'], startDate: input['startDate']});
```
or even
```
console.log(input['term'], input['interval'], input['startDate']);
```

File: trendly/src/app/app.module.ts [hunk 16]
```diff
@@ -1,32 +1,70 @@
...
-import { TopBarComponent } from './top-bar/top-bar.component';
+import {RouterModule} from '@angular/router';
+import { FormsModule, ReactiveFormsModule } from '@angular/forms';
+import { HttpClientModule } from '@angular/common/http';
+import {BrowserAnimationsModule} from '@angular/platform-browser/animations';
+// import { HashLocationStrategy, LocationStrategy } from '@angular/common';
+// {provide: LocationStrategy, useClass: HashLocationStrategy}
+
 
+
```
[the remark this answers is not in the document]
Comment (adiassulin) [reply trendly/src/app/app.module.ts:11]: Done.

File: trendly/src/app/app.module.ts [hunk 17]
```diff
@@ -1,32 +1,70 @@
 import { BrowserModule } from '@angular/platform-browser';
 import { NgModule } from '@angular/core';
-import { RouterModule } from '@angular/router'
-import { TopBarComponent } from './top-bar/top-bar.component';
+import {RouterModule} from '@angular/router';
```
[the remark this answers is not in the document]
Comment (adiassulin) [reply trendly/src/app/app.module.ts:3]: Done.

File: trendly/src/app/app.module.ts [hunk 18]
```diff
@@ -1,32 +1,70 @@
...
+import { TopBarComponent } from './top-bar/top-bar.component';
+import { HistogramyComponentComponent } from './histogramy-component/histogramy-component.component';
+import { InputsComponent } from './inputs/inputs.component';
+import { TermInputComponent } from './term-input/term-input.component';
+import { CountryInputComponent } from './country-input/country-input.component';
+import { DateInputComponent } from './date-input/date-input.component';
+import { IntervalInputComponent } from './interval-input/interval-input.component';
+import { HistogramSectionComponent } from './histogram-section/histogram-section.component';
 import { MainPageComponent } from './main-page/main-page.component';
 // import { ClusterlyComponent } from './clusterly/clusterly.component';
```
[the remark this answers is not in the document]
Comment (adiassulin) [reply trendly/src/app/app.module.ts:32]: removed.

File: trendly/src/app/data.service.ts [hunk 19]
```diff
@@ -0,0 +1,77 @@
...
+const START_DATE_PARAMETER = 'startDate=';
+const END_DATE_PARAMETER = 'endDate=';
+const COUNTRY_PARAMETER = 'country=';
+const INTERVAL_PARAMETER = 'interval=';
+const EMPTY_STR = '';
+const TYPE_STRING = 'string';
+
+/**
+ * Calls the servlets and returns the required data according to the given
+ * oarameters.
```
[the remark this answers is not in the document]
Comment (adiassulin) [reply trendly/src/app/data.service.ts:16]: changed it

File: trendly/src/app/data.service.ts [hunk 20]
```diff
@@ -0,0 +1,77 @@
...
+ */
+@Injectable({providedIn: 'root'})
+export class DataService {
+  private options = {observe: 'body' as const, responseType: 'json' as const};
```
[the remark this answers is not in the document]
Comment (adiassulin) [reply trendly/src/app/data.service.ts:20]: changed to inline.

File: trendly/src/app/data.service.ts [hunk 21]
```diff
@@ -0,0 +1,77 @@
...
+   * @param term
+   * @param startDate
+   * @param endDate
+   * @param country
+   * @param interval
+   */
+  public callServlet(
+      servletName: string, term: string|string[], startDate: string,
+      endDate: string, country: string, interval: number = 1) {
+    let basrUrl: string = servletName + START_PARAMETERS;
```
[the remark this answers is not in the document]
Comment (adiassulin) [reply trendly/src/app/data.service.ts:37]: changed to const.

File: trendly/src/app/data.service.ts [hunk 22]
```diff
@@ -0,0 +1,77 @@
...
+
+const START_PARAMETERS = '?';
+const NEXT_PARAMETER = '&';
+const TERM_PARAMETER = 'term=';
+const START_DATE_PARAMETER = 'startDate=';
+const END_DATE_PARAMETER = 'endDate=';
+const COUNTRY_PARAMETER = 'country=';
+const INTERVAL_PARAMETER = 'interval=';
+const EMPTY_STR = '';
+const TYPE_STRING = 'string';
```
[the remark this answers is not in the document]
Comment (adiassulin) [reply trendly/src/app/data.service.ts:12]: changed to inline strings.

File: trendly/src/app/data.service.ts [hunk 23]
```diff
@@ -0,0 +1,77 @@
...
+   * Calls the required servlet with the given parameters and returns the
+   * response.
+   * @param servletName
+   * @param term
+   * @param startDate
+   * @param endDate
+   * @param country
+   * @param interval
+   */
+  public callServlet(
```
[the remark this answers is not in the document]
Comment (adiassulin) [reply trendly/src/app/data.service.ts:34]: added public function for each servlet.

File: trendly/src/app/histogram-section/histogram-section.component.ts (hunk 10 above)
Comment (adiassulin) [reply trendly/src/app/histogram-section/histogram-section.component.ts:68]: removed.

File: trendly/src/app/histogram-section/histogram-section.component.ts (hunk 11 above)
Comment (adiassulin) [reply trendly/src/app/histogram-section/histogram-section.component.ts:73]: Done.

File: trendly/src/app/histogramy-component/histogramy-component.component.ts (hunk 12 above)
Comment (adiassulin) [reply trendly/src/app/histogramy-component/histogramy-component.component.ts:1]: changed it.

File: trendly/src/app/histogram-section/histogram-section.component.ts (hunk 13 above)
Comment (adiassulin) [reply trendly/src/app/histogram-section/histogram-section.component.ts:115]: cool, will modify it.

File: trendly/src/app/histogramy-component/histogramy-component.component.ts (hunk 14 above)
Comment (adiassulin) [reply trendly/src/app/histogramy-component/histogramy-component.component.ts:29]: removed it.

File: trendly/src/app/histogramy-component/histogramy-component.component.ts (hunk 15 above)
Comment (adiassulin) [reply trendly/src/app/histogramy-component/histogramy-component.component.ts:40]: Thanks, will use it.

**Review by ewebdev — APPROVED** [review]

**Pull request merged**